# Lab | Inferential statistics - T-test & P-value

## One tailed t-test 
In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file files_for_lab/machine.txt. Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as st

In [2]:
# load data with specifying the encoding type
machine = pd.read_csv('files_for_lab/machine.txt', sep="\t", encoding="utf-16")
machine

New machine      Old machine
0         42.1             42.7
1         41.0             43.6
2         41.3             43.8
3         41.8             43.3
4         42.4             42.5
5         42.8             43.5
6         43.2             43.1
7         42.3             41.7
8         41.8             44.0
9         42.7             44.1

In [3]:
machine.rename(columns = {'New machine': 'new_machine', '    Old machine': 'old_machine'}, inplace=True)

In [4]:
machine['diff'] = machine['old_machine'] - machine['new_machine']
machine.head()

new_machine  old_machine  diff
0         42.1         42.7   0.6
1         41.0         43.6   2.6
2         41.3         43.8   2.5
3         41.8         43.3   1.5
4         42.4         42.5   0.1

Null hypothesis: New machine packing time is the same as or higher than the old machine

Alternative hypothesis: New machine packing time is lower than the old machine

$$H0: \mu_{new} < \mu_{old} $$
$$H1: \mu_{new} >= \mu_{old} $$

In [11]:
s, t = st.ttest_ind(machine['new_machine'], machine['old_machine'])
if s < 0:
    one_tailed_p_value = t / 2
else:  
    one_tailed_p_value = 1 - t / 2  

print('statistic =', s, '\none-tailed pvalue =', one_tailed_p_value)

statistic = -3.3972307061176026 
one-tailed pvalue = 0.0016055712503872579


p_value is significantly lower than alpha

We reject the null hypothesis since. 

-> We can conclude that the new machine is faster 

## Matched Pairs Test 
In this challenge we will compare dependent samples of data describing our Pokemon (file files_for_lab/pokemon.csv). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.

In [17]:
pokemon = pd.read_csv('files_for_lab/pokemon.csv')
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

Null hypothesis: Defense and attack scores are equal.

Alternative hypothesis: Defense and attack scores are not equal.

$$H0: \mu_{Att} = \mu_{Def} $$
$$H1: \mu_{Att} \neq \mu_{Def} $$

In [ ]:
# Given the unequality on the alternative, we need to use a "two-sided" test.

In [18]:
s,t = st.ttest_rel(pokemon['Defense'],pokemon['Attack'])
print('statistic =', s,'\npvalue =', t)

statistic = -4.325566393330478 
pvalue = 1.7140303479358558e-05


p-value is lower than alpha (0.05), so null hypotesis is rejected. Attack and defense scores are not equal.

# Optional : Inferential statistics - ANOVA


Note: The following lab is divided in 2 sections which represent activities 3 and 4.

## Part 1

In this activity, we will look at another example. Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on:
    - Null hypothesis
    - Alternate hypothesis
    - Level of significance
    - Test statistic
    - P-value
    - F table

### Context

Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data `anova_lab_data.xlsx` file in the `files_for_lab` folder  

- State the null hypothesis
- State the alternate hypothesis
- What is the significance level
- What are the degrees of freedom of model, error terms, and total DoF

Data was collected randomly and provided to you in the table as shown: [link to the image - Data](https://education-team-2020.s3-eu-west-1.amazonaws.com/data-analytics/7.05/7.05-lab_data.png)

In [19]:
anova = pd.read_excel('files_for_lab/anova_lab_data.xlsx')
anova.head()

Power   Etching Rate
0  160 W          5.43
1  180 W          6.24
2  200 W          8.79
3  160 W          5.71
4  180 W          6.71

In [20]:
anova.groupby('Power ').agg({'Etching Rate':np.mean})

Etching Rate
Power               
160 W          5.792
180 W          6.238
200 W          8.318

In [21]:
len(anova)

15

Null hypothesis: 160 = 180 = 200

Alternative hypothesis: 160 != 180 | 160 != 200 | 180 != 200

$$H0: \mu_{1}=\mu_{2}=\mu_{3}=⋯=\mu_{n}$$
$$H1: \mu_{i}\neq\mu_{j}$$

Significance level: 0.05

Degrees of freedom: 14

## Part 2 
- In this section, use the Python to conduct ANOVA.
- What conclusions can you draw from the experiment and why?